src/main/java/uk/ac/ic/wlgitbridge/bridge/context/ContextStore.java

```diff
@@ -0,0 +1,132 @@
+ * Contexts are wrappers around a ProjectLock, and any other
+ * project-specific resources that need to be managed during
+ * an "operation" on a project.
+ *
+ * Contexts are kept in a ThreadLocal variable, so they can
+ * be re-used once they have been created.
+ *
+ * The intended use is to open a context at the 'top' of a
+ * git operation (push, pull, swap), and lock the project.
+ * Then, deeper in the call-stack the same context can be
+ * retrieved and used.
+ *
+ * (In practice, this enables us to use a Postgres connection
+ * to lock a row in the database, and hold a transaction open,
+ * to be used further down the call stack.)
+ */
+public class ContextStore {
+  private ProjectContextFactory projectContextFactory;
+  static private ContextStore instance;
+  private boolean shuttingDown;
+
+  private ThreadLocal<HashMap<String, ProjectContext>> localContexts;
+
+  private ContextStore(ProjectContextFactory contextFactory) {
+    this.projectContextFactory = contextFactory;
+    this.shuttingDown = false;
+    this.localContexts = new ThreadLocal<HashMap<String, ProjectContext>>();
+  }
+
+  /**
+   * Must be called early in application life-cycle.
+   * - @see GenericProjectContextFactory
+   * - @see PostgresProjectContextFactory
+   */
+  public static void Initialize(ProjectContextFactory contextFactory) throws ContextStoreAlreadyInitialisedException {
+    Log.info("Initialize ContextStore");
+    if (ContextStore.instance != null) {
+        Log.error("ContextStore already initialized");
+      throw new ContextStoreAlreadyInitialisedException();
+    }
+    ContextStore.instance = new ContextStore(contextFactory);
+  }
+
+  /**
+   * Reset the ContextStore, only to be used in tests,
+   * hence the __scary__ underscores
+   */
+  public static void __Reset__() {
+    Log.info("Reset ContextStore");
+    if (ContextStore.instance != null) {
+      ContextStore.Instance().stop();
+      ContextStore.instance = null;
+    }
+  }
+
+  public static ContextStore Instance()  {
+    if (ContextStore.instance == null) {
+      throw new RuntimeException("ContextStore not initialized");
+    }
+    return ContextStore.instance;
+  }
+
+  /**
+   * Cleanup contexts, and pass on the close signal.
+   * Essential for cleanly shutting down the server
+   */
+  public synchronized void stop() {
+    Log.info("Stopping ContextStore");
+    this.clearLocalContexts();
+    if (this.projectContextFactory != null) {
+      this.projectContextFactory.close();
+    }
+  }
+
+  private synchronized void ensureThreadLocalStorage() {
+    if (this.localContexts.get() == null) {
+      HashMap<String, ProjectContext> contextMap = new HashMap<String, ProjectContext>();
+      this.localContexts.set(contextMap);
+    }
+  }
+
+  private synchronized ProjectContext getFromThreadLocalStorage(String projectName) {
+    this.ensureThreadLocalStorage();
+    HashMap<String, ProjectContext> contextMap = this.localContexts.get();
```

Comment:
As mentioned above, I don't think this should be a HashMap. Instead, what if we looked something like this:

```Java
private ThreadLocal<ProjectContext> localContext;

private synchronized ProjectContext getFromThreadLocalStorage(String projectName) {
  ProjectContext context = this.localContext.get();
  if (context == null) {
    context = this.projectContextFactory.makeProjectContext(projectName);
    this.localContext.set(context);
  }
  if (context.getProjecName() != projectName) {
    throw RuntimeError("Project name mismatch in thread-local storage");
  }
  return context;
}
```

Note that in order for this to work, we have to have a mechanism to release the context (delete the thread-locally storage variable) when we are done with it. But we need to do this anyway to avoid a leak, and we can do it after we unlock.

Reply:
Discussed in call:

Actually, this is correct. We sometimes need to lock multiple projects at once, when we are migrating things around.